cpu/stm32/periph/usbdev_fs.c

```diff
@@ -0,0 +1,671 @@
+{
+    for (size_t i = 0; i < USBDEV_NUMOF; i++) {
+        _setup(&_usbdevs[i], &stm32_usbdev_fs_config[i]);
+    }
+}
+
+static void _enable_usb_clk(void)
+{
+    /* Validate USB Supply */
+    PWR->CR2 |= PWR_CR2_USV;
+    RCC->APB1SMENR1 |= RCC_APB1SMENR1_USBSMEN;
+    /* Move it to specific init */
+    RCC->CSR |= RCC_CSR_RFRSTS;
+    RCC->CSR &= ~RCC_CSR_RFRSTS;
+    /* Enable CRS with auto trim enabled */
+    CRS->CR |=  CRS_CR_AUTOTRIMEN |  CRS_CR_CEN;
+}
+
+static void _enable_gpio(const stm32_usbdev_fs_config_t *conf)
+{
+    gpio_init(conf->dp, GPIO_IN);
+    gpio_init(conf->dm, GPIO_IN);
+    /* Configure AF for the pins */
+    gpio_init_af(conf->dp, conf->af);
+    gpio_init_af(conf->dm, conf->af);
+}
+
+static void _set_ep_in_status(uint16_t *val, uint16_t mask)
+{
+    /* status endpoints bits can only be toggled, writing 0
+       as no effect. Thus, check which bits should be set
+       then XOR it with their current value */
+    if (mask & USB_EPTX_DTOG1) {
+        *val ^= USB_EPTX_DTOG1;
+    } else {
+        if (*val & USB_EPTX_DTOG1) {
+            *val |= USB_EPTX_DTOG1;
+        }
+    }
+    if (mask & USB_EPTX_DTOG2) {
+        *val ^= USB_EPTX_DTOG2;
+    } else {
+        if (*val & USB_EPTX_DTOG2) {
+            *val |= USB_EPTX_DTOG2;
+        }
+    }
+}
+
+static void _set_ep_out_status(uint16_t *val, uint16_t mask)
+{
+       /* status endpoints bits can only be toggled, writing 0
+       as no effect. Thus, check which bits should be set
+       then XOR it with their current value */
+    if (mask & USB_EPRX_DTOG1) {
+        *val ^= USB_EPRX_DTOG1;
+    } else {
+        if (*val & USB_EPRX_DTOG1) {
+            *val |= USB_EPRX_DTOG1;
+        }
+    }
+    if (mask & USB_EPRX_DTOG2) {
+        *val ^= USB_EPRX_DTOG2;
+    } else {
+        if (*val & USB_EPRX_DTOG2) {
+            *val |= USB_EPRX_DTOG2;
+        }
+    }
+}
+
+static inline void _usb_attach(stm32_usbdev_fs_t *usbdev)
+{
+    /* Some ST USB IP doesn't have this feature */
+#ifdef USB_BCDR_DPPU
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    bool syncing = false;
+    /* Enable DP pullup to signal connection */
+    _global_regs(conf)->BCDR |= USB_BCDR_DPPU;
+    do {
+        syncing = CRS->ISR & 0x00000008;
+    } while (!syncing);
+#else
+    (void)usbdev;
+#endif /* USB_BCDR_DPPU */
+}
+
+static inline void _usb_detach(stm32_usbdev_fs_t *usbdev)
+{
+    /* Some ST USB IP doesn't have this feature */
+#ifdef USB_BCDR_DPPU
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    DEBUG_PUTS("usbdev_fs: Detaching from host");
+
+    /* Disable DP pullup to signal disconnection */
+    CLRBIT(_global_regs(conf)->BCDR, USB_BCDR_DPPU);
+#else
+    (void)usbdev;
+#endif
+}
+
+static void _set_address(stm32_usbdev_fs_t *usbdev, uint8_t address)
+{
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    _global_regs(conf)->DADDR = USB_DADDR_EF | (address & USB_DADDR_ADD);
+}
+
+void isr_usb_hp(void) {
+
+}
```

Comment:
Is this isr triggered by the hardware? If not, maybe we can leave it to the default handler. Or maybe add a debug print here

Reply:
Yup, this is a leftover. This ISR isn't used right now.

Reply:
Removed.